nroff/nroffdoc/difftags.1
.Id $Id: difftags.1,v 1.1 2000-10-21 14:31:32 jon Exp $
.Im mantools utility
.TH difftags 1
.XI difftags - "Analyse and compare an Nroff tags file and C tags file."
.SH NAME
difftags \- Analyse and compare an Nroff tags file and C tags file.
.SH SYNOPSIS
.sp 1
.na
\fBdifftags\fR \fI<cfile1>\fR \fI<docfile2>\fR \fC[\fR\fI<outFile>\fR\fC]\fR
.ad
.SH DESCRIPTION
\fBdifftags\fR is a utility used to compare tag files and is generally used to
compare a program tag file (generated by
.Hr ctags 1 )
with a Nroff tags file generated by
.Ht ntags 1 .
The  utility  generates a  difference  file which  annotates  the  differences
between  the tag  files.  The  utility  is only  interested  in the tag  names
determining which symbol definitions are and are not defined.
.PP
The command  line  accepts  two tag files, a code tag file  \fI<cfile1>\fR  is
specified as the first tag file (typically generated by
.Hr ctags 1 "),"
the document tag file \fI<docfile2>\fR is specified second (generated by
.Ht ntags 1 ")."
An optional  output file may be  specified  as the third  argument, if this is
omitted then the differences are generated on \fIstdout\fR.
.PP
The generated difference file provides information as to:-
.BS
The labels that have been  documented  e.g. match in names between the two tag
files.
.BU
The labels that are not documented e.g. a label appearing in the code tag file
and not the nroff tag file.
.BU
The labels that are \fIover\fR  documented e.g. a label appearing in the nroff
tag file and not in the code file.
.BE
The generated output file takes the following form:-
.CS
.ne 6
Files:      ..\\src\\tags [0]          tags [0]

** Undef    VTXFloat32Write [1]      ---
** Undef    VTXFloat64Write [2]      ---
** Undef    VTX_DBPALSAVE [3]        ---
   ....
   ....
** Undef    VTXtestForVTX [88]       ---
** Undef    VTXtextureWrite [89]     ---
** Undef    VertexInList [90]        ---
-- ok       byte2_flip [91]          byte2_flip [1]
-- ok       byte4_flip [92]          byte4_flip [2]
-- ok       byte8_flip [93]          byte8_flip [3]
** Unknown  ---                      dfl [4]
-- ok       dflBackupName [94]       dflBackupName [5]
   ....
   ....
-- ok       dflCreateVtxName [104]   dflCreateVtxName [15]
-- ok       dflError [105]           dflError [16]
** Unknown  ---                      dflFILE [17]
** Unknown  ---                      dflFILEPTR [18]
-- ok       dflFormatName [106]      dflFormatName [19]
   ....
   ....
** Undef    dvfUnlinkObject [878]    ---
** Undef    dvfWriteTab [879]        ---
** Unknown  ---                      progname [500]

Total:      ..\\src\\tags [880]        tags [501]

Matched = 402. Undefined = 477. Unknown = 98
.CE
A three column  output is  generated,  column 1 provides  status  information,
columns 2 and 3  contain  the  labels  taken  from the tags  files,  each name
includes the source line number in square brackets.
.PP
The status column shows the status of the name comparison as follows:-
.BP "** Undef"
The name exists in the code tags file, but not in the nroff tags file.
.BP "** Unknown"
The name exists in the nroff tags file, but not in the code tags file.
.BP "-- ok"
The name has been defined in both the nroff and code tags files.
.PP
The end of the difference  file  identifies  the number of tags that have been
matched and totals the number of tags that are undefined or unknown.
.SH USAGE
The  \fBdifftags\fR  utility is primarily  intended to enable a coding  module
symbols to be checked against the documentation files to ensure that they have
been  documented. The following  steps are generally taken when  documenting a
module.
.BS
Execute
.Hr ctags 1
on the source files.
.BU
Copy the  generated  tags file and remove  all  internal  labels  that are not
documented, leaving all of the entry points that MUST be documented.
.BU
Execute
.Ht ntags 1
on the documentation files.
.BU
Using the  edited  code tags file and nroff  tags file use  \fBdifftags\fR  to
determine the difference.
.BU
Iteratively correct the documentation, re-run
.Ht ntags 1
and \fBdifftags\fR until all symbols have been resolved properly.
.BE 0
.SH "SEE ALSO"
.na
.Hr ctags 1 ,
.Ht ntags 1 .
.ad
.FH


nroff/nroff/difftags.c
/****************************************************************************
 *
 *			   Copyright 1995 Jon Green.
 *			      All Rights Reserved
 *
 *
 *  System        :
 *  Module        :
 *  Object Name   : $RCSfile: difftags.c,v $
 *  Revision      : $Revision: 1.1 $
 *  Date          : $Date: 2000-10-21 14:31:25 $
 *  Author        : $Author: jon $
 *  Last Modified : <030597.1443>
 *
 *  Description
 *
 *  Notes
 *
 *  History
 *
 *  $Log: not supported by cvs2svn $
 *
 ****************************************************************************
 *
 *  Copyright (c) 1995 Jon Green.
 *
 *  All Rights Reserved.
 *
 *  This Document may not, in whole or in part, be copied,
 *  photocopied, reproduced, translated, or reduced to any
 *  electronic medium or machine readable form without prior
 *  written consent from Jon Green.
 ****************************************************************************/

static const char rcsid[] = "@(#) : $Id: difftags.c,v 1.1 2000-10-21 14:31:25 jon Exp $";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include <utils.h>

FILE *f1;                               /* First input file */
FILE *f2;                               /* Second input file */
FILE *fo;                               /* Output file */

static char b1 [1024];
static char b2 [1024];
static int  lineNo1 = 0;
static int  lineNo2 = 0;
static int  undefinedCount = 0;
static int  matchCount = 0;
static int  unknownCount = 0;

void
printLine (char *cause, char *s1, int l1, char *s2, int l2)
{
    int i;
    char buffer [1024];

    fprintf (fo, "%s  ", cause);
    if (s1 != NULL) {
        sprintf (buffer, "%s [%d]", s1, l1);
        fprintf (fo, "%s", buffer);
        i = strlen (buffer);
    }
    else
    {
        fprintf (fo, "---");
        i = 3;
    }

    while (i < 55) {
        fprintf (fo, " ");
        i++;
    }
    if (s2 != NULL)
        fprintf (fo, "%s [%d]\n", s2, l2);
    else
        fprintf (fo, "---\n");
}

void
getLine (FILE **fp, char *buffer, int *lineNo)
{
    int  status;
    char junk [1024];

    if (*fp == NULL) {
        *buffer = '\0';
        return;
    }
    status = fscanf (*fp, "%[^\t]\t%[^\n]\n", buffer, junk);
    if (/*(status == EOF) ||*/ (status != 2)) {
        fclose (*fp);
        *fp = NULL;
        *buffer = '\0';
    }
    *lineNo += 1;
    return;
}

int
main (int argc, char *argv [])
{
    int status;

    if ((argc < 3) || (argc > 4)) {
        fprintf (stdout, "Usage: difftags <cfile1> <docfile2> [<fileout>]\n");
        exit (2);
    }

    /* Process each of the files in turn */

    if ((f1 = fopen (argv [1], "r")) == NULL) {
        fprintf (stderr, "Cannot open file [%s]\n", argv[1]);
        exit (2);
    }

    if ((f2 = fopen (argv [2], "r")) == NULL) {
        fprintf (stderr, "Cannot open file [%s]\n", argv[2]);
        exit (2);
    }

    if (argc == 4) {
        if ((fo = fopen (argv [3], "w")) == NULL) {
            fprintf (stderr, "Cannot open output file [%s]\n", argv[3]);
        exit (2);
        }
    }
    else
        fo = stdout;

    getLine (&f1, b1, &lineNo1);
    getLine (&f2, b2, &lineNo2);

    printLine ("Files:    ", argv [1], 0, argv [2], 0);
    fprintf (fo, "\n");
    while ((f1 != NULL) || (f2 != NULL))
    {
        /* Evaluate the matched status */

        if (b1[0] == '\0')
            status = 1;
        else if (b2[0] == '\0')
            status = -1;
        else
            status = strcmp (b1, b2);

        /* Given the status print the result. */

        if (status == 0) {
            printLine ("-- ok     ", b1, lineNo1, b2, lineNo2);
            getLine (&f1, b1, &lineNo1);
            getLine (&f2, b2, &lineNo2);
            matchCount++;
        }
        else if (status < 0) {
            printLine ("** Undef  ", b1, lineNo1, NULL, 0);
            getLine (&f1, b1, &lineNo1);
            undefinedCount++;
        }
        else {
            printLine ("** Unknown", NULL, 0, b2, lineNo2);
            getLine (&f2, b2, &lineNo2);
            unknownCount++;
        }
    }
    fprintf (fo, "\n");
    printLine ("Total:    ", argv [1], lineNo1, argv [2], lineNo2);
    fprintf (fo, "\n");
    fprintf (fo, "Matched = %d. Undefined = %d. Unknown = %d\n",
             matchCount, undefinedCount, unknownCount);
    fclose (fo);
    return (0);
}
